### ISSUE TITLE
Introduce a way to detect stateful event args

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#64]
Event args and action args vs stateful args were discussed at some length in #15. That pretty much brought us to the current state of the code, although since that time our thinking about the technical differences and importance of stateful vs non-stateful args has evolved somewhat. Much discussion related to this took place in #59.

The conclusion there was that we could identify repeatable events by the fact that they had only stateful args, if only we had a way to easily identify in the API whether an arg was stateful or not.

Another place where this would be useful is in determining whether a reaction can be fired retroactively: if it references any stateful args, it cannot. See #1.

So it seems it is worth the time and effort to provide some way of identifying stateful args. Exactly how that should be done is what this issue will explore.

**Comment by JDGrimes** [comment]
## Aliases

Something that I've wondered before is whether there is a one-to-one correlation between stateful args and aliased args. (See https://github.com/WordPoints/hooks-api/issues/15#issuecomment-145143512 and https://github.com/WordPoints/hooks-api/issues/15#issuecomment-145249499).

This might make sense, but it isn't necessarily true just by nature. It would be an arbitrary constraint we'd have to place on arg naming. Perhaps it is true that all alias args are stateful, but not all stateful args _have_ to be aliased (unless just because we say so). And I'm not even sure that all alias args are stateful.

So let's ask the question of why we need aliases. The answer is because we can't otherwise have two args of the same type for an event; two `user` entities for example. So we'd call one arg `user` and the other `another:user` (or whatever would be appropriate). This allows us to continue do identify the entity slug from the arg slug. 

The only time that we'd ever have two different entities of the same type would be if they weren't related to each other. Because if they were related, we'd only need to include the arg for one of them and the other one would be accessible through the relationship. And if the two entities aren't related, then the event must be repeatable, and therefore both entities must be stateful.

Actually, I suppose it is possible that although the two entities were related, the relationship might be tenuous enough that it would be helpful to the user to have both entities available as event args. For example, consider the event for when a user becomes a member of a site on the network. We'd provide both the user and site args for the event, because although the user and site would then be related, there would be no easy way to target either of them specifically through those relationships, because the relationships would be through entity arrays.

Otherwise we'd basically have just proven that an event can never have more than one arg unless it is a repeatable event. Yet this isn't true. Instead, a non-repeatable event can have more than one arg, but only if each of the args is related to another one of the args through a (one-to-many) relationship.

So, since it is possible for non-repeatable events to have multiple args, and since these args will be non-stateful by definition, and since it is possible that more than one of those args may be of the same type:

So we conclude that aliases can be used for non-stateful entities.

Also, just because a repeatable event are stateful doesn't mean that any of the really have to be aliased, unless there are multiple ones of the same type (but then it is only necessary to alias all but one of them).

### Conclusion

So there really is no correlation between stateful and aliased args, although we might suggest, for the sake of simplicity, that only args which really have to be aliased should be. I suppose though, that when there are multiple args of the same type it might be useful to alias all of them, since the unaliased one, when considered on its own, might be ambiguous.

**Comment by JDGrimes** [comment]
This leaves us with two options:
- Introducing an interface by which to identify stateful args as originally suggested in #15.
- Storing the data on whether each arg is stateful or not in memory somewhere.

## Interface

One benefit of an interface is that it allows us to easily determine whether the arg is stateful just using the arg object. That is also a weakness though, because it means you must have the arg object at hand to determine this. 

Another benefit is that it allows us to specify special methods that stateful args should have. I'm not sure that we need any at the moment, so perhaps that is a moot point. Although #15 did suggest a `set_the_value_from_state` method (for "stateful entities" though).

## Data

Storing this information in the class registry or somewhere like that would mean that it would be possible to get the info on whether an arg was stateful using just the arg slug. We might also be able to get this data for all args of an event at once.

The downside is that now we have to go and look up that info somewhere other than the entity object itself, which can be a pain when you already have the entity object. Of course, we could always provide helper methods for that in the entity object, or even construct the object with this data to start with.

Perhaps we'd use both this and an interface together to provide the best of both worlds.

**Comment by JDGrimes** [comment]
Another possibility would be to add an `is_stateful()` method to the main arg object. Then each child class could decide how best to determine whether the arg was stateful or not. It could return `true` or `false` explicitly, or allow that information to be passed into it via the constructor, or specified as a property or even an interface. Since this is going to become an integral part of this API, I think that makes sense.

**Comment by JDGrimes** [comment]
I've decided to go ahead and do this with the method. It isn't a perfect solution since it is defined per arg object, while statefulness is really something that must be determined per event (each arg object can be used for several events). This is something that we can expand upon in the future if needed.

**Comment by JDGrimes** [comment]
We might want to revisit this after #78. The determination from there is that events can only have one "signature" or primary arg. The rest are considered "stateful", but I'm not sure whether we might want to split them up into different categories. Anyway, we at least ought to change the `get_non_stateful_args()` method to `get_primary_arg()` or something like that.

**Comment by JDGrimes** [comment]
> Anyway, we at least ought to change the get_non_stateful_args() method to get_primary_arg() or something like that.

So let's work on that first.

In order to do this we need to be able to identify the primary or "signature" arg. Right now that is just implicit—we could check if the arg was stateful, and it it isn't it must be the primary arg. But it seems like we should change that to be more explicit. I guess that should be for another ticket though.

**Comment by JDGrimes** [comment]
I think anything more we need to do here should be in new tickets.
